Integrating FairXplainer with AIF360

bishwamittra opened the pull request:
Two files are added to integrate FairXplainer. 

1. aif360/sklearn/explainers/bias_explainer.py
2. examples/sklearn/demo_bias_explaination.ipynb 

(1) contains the interface to call FairXplainer from AIF360 and (2) provides a notebook demonstration.

mnagired reviewed (commented):
Hi @bishwamittra,

Thank you for contributing FairXplainer!

I briefly went through the demo notebook and had a quick suggestion: 

Could you please provide a few sentences for each of the (3) different explanations (statistical parity, equalized odds, and predictive parity)? Specifically, a sentence or two after each of the plots, explaining how to interpret / contextualize the plots might be helpful for new users (especially folks unfamiliar with SHAP-related plots).

Other than that, the code and notebooks look ok to me! Thanks a lot :)

cc: @hoffmansc

bishwamittra commented:
Hi @mnagired, 

I have added descriptions of explanations according to your suggestion. Hope this addresses your question. 

Best regards,
Bishwa

mnagired commented:
Thanks! Yes, this looks good to me.

mnagired reviewed (commented):
Looks good to me, thanks for updating!

mnagired commented:
@hoffmansc, this PR looks good to me. Please have a look whenever you get a chance :)

hoffmansc commented on aif360/sklearn/explainers/bias_explainer.py:
```diff
@@ -0,0 +1,193 @@
+
+    Args:
+        clf: the trained model to explain
+            clf.predict() should be implemented
```
does it matter if this doesn't return 0/1? what if 0 is favorable?

bishwamittra replied:
The code is implemented in a way that 1 is the favourable class. Also, I assume discrete (particularly Boolean) prediction by the classifier.

hoffmansc commented on aif360/sklearn/explainers/bias_explainer.py:
```diff
@@ -0,0 +1,193 @@
+        clf: the trained model to explain
+            clf.predict() should be implemented
+        X (pandas.DataFrame): the dataset (containing the features) to measure the bias on
+        sensitive_features (list): list of sensitive attributes
```
can this be renamed `prot_attr` for consistency with metrics, etc.?

bishwamittra replied:
Sure. I am addressing this.

hoffmansc reviewed (changes requested):
Hi, @bishwamittra. Thanks for all your hard work! It's looking good, I just had a couple comments.

If you could also include a small unit test file just to catch any errors that might come up due to changes in the future, that would be great.

bishwamittra commented on aif360/sklearn/explainers/bias_explainer.py:
```diff
@@ -0,0 +1,193 @@
+import numpy as np
+
+
+def explain_statistical_parity(
```
I am changing the method name accordingly.

bishwamittra commented on aif360/sklearn/explainers/bias_explainer.py:
```diff
@@ -0,0 +1,193 @@
+        maxorder=2, 
+        spline_intervals=5,
+        k=7,
+        seed=22, 
```
Yes, that is possible.

bishwamittra commented on aif360/sklearn/explainers/bias_explainer.py:
```diff
@@ -0,0 +1,193 @@
+        verbose (bool): whether to print the results
+
+    Returns:
+        pandas.DataFrame : the bias weights
```
This is indeed possible. Let me address this.

bishwamittra commented on aif360/sklearn/explainers/bias_explainer.py:
```diff
@@ -0,0 +1,193 @@
+        verbose (bool): whether to print the results
+
+    Returns:
+        pandas.DataFrame : the bias weights as list for y_true = 0 and y_true = 1
```
Thanks for the correction.

bishwamittra commented on aif360/sklearn/explainers/bias_explainer.py:
```diff
@@ -0,0 +1,193 @@
+
+    Args:
+        result (pandas.DataFrame): the bias weights
+        kwargs: arguments to pass to the plot function
```
Yes. I am modifying the arguments.

bishwamittra commented:
Hi @hoffmansc 

I have addressed your comments. The summary of changes are:

1. pytest added for bias explanation,
2. fairxplainer returns explanation results without bias value. Bias value can be computed by existing metric in aif360, as demonstrated in the notebook. However, I have not been able to compute predictive parity (sufficiency metric) using aif360. Therefore, I have added an optional argument "return_bias", which is by default False. If "return_bias=True", the bias computed by FairXplainer is additionally returned. 
3. Other comments are addressed to my best efforts.

Thanks,
Bishwa

hoffmansc reviewed (changes requested):
Hi, Bishwa. Thanks for your updates.

Regarding the predictive parity metric, I believe you should be able to compute it just like equalized odds but using `base_rate` instead of `selection_rate` and swapping `y_train == y` for `clf.predict(X_train) == y`, i.e.,
```py
intersection(base_rate, y_train[clf.predict(X_train) == y], prot_attr=prot_attr)
```
but in any case, I’m fine keeping the `return_bias` option for now. I originally thought the metrics could be directly computed with `statistical_parity_difference`, `average_odds_difference`, and `average_predictive_value_difference` from `aif360.sklearn.metrics` but I see now it’s slightly different. Eventually those should functions be generalized to return the max differences (perhaps with PR #413) but that’s out of scope for this PR.

Another thing I noticed as I dug deeper is the `FairXplainer` expects the sensitive features to be columns from the dataset and y labels need to be 0/1. Is it possible to pass an explicit set of groups from [`check_groups`](https://github.com/Trusted-AI/AIF360/blob/master/aif360/sklearn/utils.py#L44) to be more in line with other aif360 metrics? This way you could have, e.g., a binary age sensitive attribute but continuous age feature. Also, if y has to be binary, can we have a `pos_label` arg and do the conversion in the function? Is it necessary to pass the classifier or can we just always pass the predictions and binarize those?

Other than that, a couple small things are still missing:
- [ ] `fairxplainer` needs to be added to [setup.py](https://github.com/Trusted-AI/AIF360/blob/master/setup.py#L28) under “extras”, i.e., `’FairXplainer’: [‘fairxplainer’],` as well as requirements.txt
	- [ ] in order for pip to install the relevant dependencies automatically, you’ll need to also add the top-level requirements to `install_requires` in setup.py of the upstream fairxplainer package as well. Only the “leaves” of the dependency tree need to be added, e.g., shap, feature-engine, etc.
- [ ] there needs to be a new section added to the [docs page](https://github.com/Trusted-AI/AIF360/blob/master/docs/source/modules/sklearn.rst?plain=1#L138) It should look like this:
```rst
:mod:`aif360.sklearn.explainers`: Fairness metric explainers
============================================================

.. automodule:: aif360.sklearn.explainers
   :no-members:
   :no-inherited-members:

FairXplainer
------------
.. currentmodule:: aif360.sklearn

.. autosummary::
   :toctree: generated/
   :template: base.rst

   explainers.fairxplain_statistical_parity
   explainers.fairxplain_equalized_odds
   explainers.fairxplain_predictive_parity
   explainers.fif_plot
```
